post/kafka testing (#18)
sysco-middleware/sysco-middleware.github.io: _posts/2019-06-24-testing-kafka-based-applications.md

---
layout: post
title: Testing Kafka based applications
categories: automatic tests
tags: [kafka, RestAssured, WireMock, test, Testcontainers, docker, junit, end-to-end]
author: NikitaZhevnitskiy
---

# Testing Kafka based applications
![streams](/images/2019-06-24-testing-kafka-based-applications/streams.jpg)

> Github: [https://github.com/sysco-middleware/kafka-testing](https://github.com/sysco-middleware/kafka-testing)

This blog is of continuation of discussion in these posts:
1. [Getting Your Feet Wet with Stream Processing - Part 2: Testing Your Streaming Application](https://www.confluent.io/blog/stream-processing-part-2-testing-your-streaming-application)
2. [Testing Event-Driven Systems](https://www.confluent.io/blog/testing-event-driven-systems)

Main purpose of this article is to share experience of testing Apache Kafka based applications and data pipelines, using two different approaches.

Keywords in use:
* **Kafka based application** - by Kafka based application I understand any application that uses Kafka API and communicates with Kafka cluster.
* **Data pipeline** - is a set of Kafka based applications that are connected into a single context.

> Examples are built using java and docker. For detailed information, check this repository on [github](https://github.com/sysco-middleware/kafka-testing).

Reviewer: Why (K)afka but (j)ava and (d)ocker?

Author: done

Reviewer 2: Is this resolved? I'm still seen java and docker with lowercase, as well as github. 

```suggestion
Examples are built using Java and Docker. For detailed information, check this repository on [GitHub](https://github.com/sysco-middleware/kafka-testing).
```

## Agenda
1. Intro

> 2. Unit tests of Kafka Streams application with [kafka-streams-test-utils](https://kafka.apache.org/20/documentation/streams/developer-guide/testing.html)

Reviewer 2: nit:
```suggestion
2. Unit tests of Kafka Streams application with [kafka-streams-test-utils](https://kafka.apache.org/documentation/streams/developer-guide/testing.html)
```

3. Integration tests with [EmbeddedKafkaCluster](https://github.com/apache/kafka/tree/trunk/streams/src/test/java/org/apache/kafka/streams/integration/utils)
4. End-to-end tests with [Testcontainers](https://www.testcontainers.org)

### 1. Intro

> Many have faced a challenge to test applications that communicate with Kafka cluster. The common issues are how to write unit test, how to automate test for Kafka application in isolation and how to handle end-to-end and system test for whole data pipeline.

Reviewer: probably plural form "tests" would fit better here

Author: done

> When I write this article, there are several approaches, some of them not resolved yet:

Reviewer 2: Rephrase this to "When I wrote this article", or "When this article was written, there were..."

Reviewer 2: nit: Also, "some of them still unresolved"

> 1. [KIP-139: Kafka TestKit library](https://cwiki.apache.org/confluence/display/KAFKA/KIP-139%3A+Kafka+TestKit+library) is still open. Once it will be ready, kafka dev community will have a proper test-kit.

Reviewer 2: ```suggestion
1. [KIP-139: Kafka TestKit library](https://cwiki.apache.org/confluence/display/KAFKA/KIP-139%3A+Kafka+TestKit+library) is still open. Once it will be ready, the Kafka community will have a proper test-kit.
```

> 2. [#26 Encapsulate EmbeddedSingleNodeKafkaCluster in a seperately-available maven/gradle/sbt](https://github.com/confluentinc/kafka-streams-examples/issues/26) dep. Kafka cluster in memory. When this post is published, there is an open issue for having release of embedded kafka cluster. [Yeva Byzek](https://www.confluent.io/blog/author/yeva-byzek/) offers 

Reviewer 2: ```suggestion
2. [#26 Encapsulate EmbeddedSingleNodeKafkaCluster in a seperately-available maven/gradle/sbt](https://github.com/confluentinc/kafka-streams-examples/issues/26) dep. Kafka cluster in memory. When this post was published, there is an open issue for having release of embedded Kafka cluster. [Yeva Byzek](https://www.confluent.io/blog/author/yeva-byzek/) offers 
```

> to use [KafkaEmbeddedCluster](https://github.com/apache/kafka/tree/trunk/streams/src/test/java/org/apache/kafka/streams/integration/utils) in this [blog-post](https://www.confluent.io/blog/stream-processing-part-2-testing-your-streaming-application). There is no existing release for these utils. I decided to duplicate this part of code base and to build an [internal module](https://github.com/sysco-middleware/kafka-testing/tree/master/embedded-cluster) as an example.

Reviewer 2: ```suggestion
to use [KafkaEmbeddedCluster](https://github.com/apache/kafka/tree/trunk/streams/src/test/java/org/apache/kafka/streams/integration/utils) in this [blog-post](https://www.confluent.io/blog/stream-processing-part-2-testing-your-streaming-application). There is no existing release for these libraries. I decided to duplicate this part of code base and to build an [internal module](https://github.com/sysco-middleware/kafka-testing/tree/master/embedded-cluster) as an example.
```

3. [KIP-247 Add public test utils for Kafka Streams](https://cwiki.apache.org/confluence/display/KAFKA/KIP-247%3A+Add+public+test+utils+for+Kafka+Streams). This KIP is ready and has release for unit testing kafka streams topologies - [kafka-streams-test-utils](https://kafka.apache.org/20/documentation/streams/developer-guide/testing.html).
4. [Testcontainers](https://www.testcontainers.org) - testing with docker containers.

### 2. Unit tests of Kafka Streams application with kafka-streams-test-utils

> [Kafka-streams-test-utils](https://kafka.apache.org/21/documentation/streams/developer-guide/testing.html) is a test-kit for testing stream topologies in memory without need to run Kafka cluster. Test-kit supports stateful operations and avro serialisation/deserialisation via mocking of schema registry.

Reviewer 2: nit:
```suggestion
[Kafka-streams-test-utils](https://kafka.apache.org/21/documentation/streams/developer-guide/testing.html) is a test-kit for testing stream topologies in memory without needing a Kafka cluster. Test-kit supports stateful operations and Avro serialization/deserialization via mocking of [Schema Registry](https://docs.confluent.io/current/schema-registry/index.html).
```

> All that you need is TopologyTestDriver, which you can init for each test with topology under test.

Reviewer 2: ```suggestion
All that is required is to have a `TopologyTestDriver`, which can be initialize on each test case where a topology is under test.
```

```java
private TopologyTestDriver testDriver;
private MockSchemaRegistryClient schemaRegistryClient;

@Before
public void start() {
    schemaRegistryClient = new MockSchemaRegistryClient();
}

@After
public void tearDown() {
    Optional.ofNullable(testDriver).ifPresent(TopologyTestDriver::close);
    testDriver = null;
    properties = null;
}
```

> I choose for testing topology with stateful operations and avro.

Reviewer 2: ```suggestion
 I choose for testing topology with stateful operations and Avro. 
```

            topicIn,
            "1",
            Person.newBuilder()
                .setId("id-1")
                .setName("nikita")
                .setLastname("zhevnitskiy")
                .build()
        );
    
    final ConsumerRecord<byte[], byte[]> inRecord2 =
        factory.create(
            topicIn,
            "2",
            Person.newBuilder()
                .setId("id-2")
                .setName("nikita")
                .setLastname("moscow")
                .build()
        );
    
    /** Act */
    testDriver.pipeInput(Arrays.asList(inRecord1, inRecord2));
    final KeyValueStore<String, Long> keyValueStore = testDriver.getKeyValueStore(storeName);
    final Long amountOfRecordWithSameName = keyValueStore.get("nikita");
    
    /** Assert */
    assertEquals(Long.valueOf(2), amountOfRecordWithSameName);
}
```

You can find full test example [here](https://github.com/sysco-middleware/kafka-testing/blob/master/streams-client/src/test/java/no/sysco/testing/kafka/streams/topology/StreamProcessingAvroTest.java) and more examples in [Apache Kafka documentation](https://kafka.apache.org/21/documentation/streams/developer-guide/testing.html). 

> Pay attention to issue [#877 - Passing Schema Registry URL twice to instantiate KafkaAvroSerializer](https://github.com/confluentinc/schema-registry/issues/877).

Reviewer 2: Might be interesting to add a bit of context about this suggestion, like why to pay attention?

Author: I think context is in issue title. For more info, here is link. I would avoid to put context because it is not related to blog post topic

### 3. Integration tests with [EmbeddedKafkaCluster](https://github.com/apache/kafka/tree/trunk/streams/src/test/java/org/apache/kafka/streams/integration/utils)
Integration test is a way of how to test services in isolation but with required dependencies. 
Embedded Kafka cluster combines broker, zookeeper and schema-registry in one. 

> To test asynchronous system I choose [awaitility library](http://www.awaitility.org). 

Reviewer 2: ```suggestion
To test asynchronous APIs I've chosen [awaitility library](http://www.awaitility.org).
```

  public void testAvroConsumer() throws InterruptedException, 
    ExecutionException, TimeoutException {

    String topic = "topic2";
    CLUSTER.createTopic(topic);

    final KafkaProducer<String, Event> producer = 
      new KafkaProducer<>(getProducerProperties());
    
    producer
      .send(
        new ProducerRecord<>(
          topic,
          "id-3",
          Event.newBuilder()
          .setId("id-3")
          .setType("type-3")
          .setContext("context-3")
          .build()))
      .get(1, TimeUnit.SECONDS);

    final KafkaConsumer<String, Event> consumer = 
      new KafkaConsumer<>(getConsumerProperties());
    consumer.subscribe(Collections.singletonList(topic));

    final ArrayList<Event> events = new ArrayList<>();

    await()
      .atMost(15, TimeUnit.SECONDS)
      .untilAsserted(() -> {
          final ConsumerRecords<String, Event> records = 
            consumer.poll(Duration.ofSeconds(1));
          for (final ConsumerRecord<String, Event> record : records) 
            events.add(record.value());
  
          assertEquals(1, events.size());
      });
  }
}
```

### 4. End-to-end tests with [Testcontainers](https://www.testcontainers.org)

> In context of Kafka based applications, end-to-end testing will be applied to  data pipelines to ensure that, first, the data integrity is maintained between applications and, second, data pipelines behave as expected. 

Reviewer 2: ```suggestion
In context of Kafka-based applications, end-to-end testing will be applied to data pipelines to ensure that, first, the data integrity is maintained between applications and, second, data pipelines behave as expected.
```
nit: replace others `%s/Kafka based/Kafka-based/g`

Testing requires all integrated components of application to be up and running - in order to be able to test them with different scenarios.
![data pipeline](/images/2019-06-24-testing-kafka-based-applications/data_pipeline.png)

> Requirements:

Reviewer 2: Upper case for software names below

* zookeeper 
* kafka broker 
* confluent schema registry 
* http-producer: REST producer 
* http-materializer: streams application, that downstreams topic data and RPC to another REST service
* db-mock: REST service and db in memory (use .json file as persistent storage), all-in-one

> I describe two approaches of how to prepare container environment for testing. Both of them require a certain start-up sequence, namely: start kafka cluster first, wait until it is ready to serve data, create required topics, register schemas and then start application services.

Reviewer 2: ```suggestion
I describe two approaches on how to prepare a container environment for testing. Both of them require a certain start-up sequence, namely: start a Kafka cluster first, wait until it is ready to receive requests, create required topics, register schemas and then run application services.
```

Testcontainers support `Waiting strategies` until containers are ready for test. Potential improvement could be implementation of health checks for each component in data pipeline.

```java
/**
 * Base test  abstract class contains setup to
 * initialize testing environment.
 */
public abstract class ContainersTestBaseIT {

    // singleton containers https://www.testcontainers.org/test_framework_integration/manual_lifecycle_control/#singleton-containers
    static  {
        final Network commonNetwork = Network.newNetwork();
        setZookeeperAndKafka(commonNetwork);
        setSchemaRegistry(commonNetwork);
        setJsonServer(commonNetwork);
        setHttProducer(commonNetwork);
        setHttpMaterializer(commonNetwork);
    }
    
    // . . . rest code
}
```

To make containers interact with each other, you should put them with the same network. 
If one container depends on another container, you might need network alias to setup communication between them. 
You can provide your own network alias for container or use an already existing one. 
Aliases can be injected as environmental variable.

> Container ports can be mapped to localhost while testing, if container's port is exposed. 

Reviewer 2: ```suggestion
Container ports can be mapped to `localhost` while testing, if container's port is exposed. 
```

It can be useful for creating a kafka topic before running streams application.

```java
private static void createTopic(String topicName) {
    // kafka container uses with embedded zookeeper
    // confluent platform and Kafka compatibility 5.1.x <-> kafka 2.1.x
    // kafka 2.1.x require option --zookeeper, later versions use --bootstrap-servers instead
    String createTopic =
        String.format(
            "/usr/bin/kafka-topics --create --zookeeper localhost:2181 --replication-factor 1 --partitions 1 --topic %s",
            topicName);
    try {
      final Container.ExecResult execResult = kafka.execInContainer("/bin/sh", "-c", createTopic);
      if (execResult.getExitCode() != 0) fail();
    } catch (Exception e) {
      e.printStackTrace();
      fail();
    }
}
```

#### Approach 2: Define containers in docker-compose file
Another approach is to prepare containers environment for testing is to define all the environment information in a docker-compose.yml. In this case Testcontainers manage composition of containers. All services share common network. Name of service uses as a network alias. This approach also supports waiting strategies. 

> Example of how to create topic in a docker-compose file can be found in [confluent examples](https://github.com/confluentinc/examples/blob/f854ac008952346ceaba0d1f1a352788f0572c74/microservices-orders/docker-compose.yml#L182-L215).

Reviewer 2: ```suggestion
Example of how to create topic in a `docker-compose` file can be found in [Confluent examples](https://github.com/confluentinc/examples/blob/f854ac008952346ceaba0d1f1a352788f0572c74/microservices-orders/docker-compose.yml#L182-L215).
```

        .withExposedService("db-mock_1", 80)
        .withExposedService("http-producer_1", 8080);
```

Test
```java
@Test
public void test_data_pipeline_flow_successful() {
    String id = UUID.randomUUID().toString();
    String from = UUID.randomUUID().toString();
    String to = UUID.randomUUID().toString();
    String text = UUID.randomUUID().toString();

    MessageJsonRepresentation messageJsonRepresentation =
        new MessageJsonRepresentation(id, from, to, text);

    RestAssured.given()
        .contentType(ContentType.JSON)
        .body(messageJsonRepresentation)
        .post(HTTP_PRODUCER_EXPOSED + "/messages")
        .then()
        .statusCode(202);

    await()
        .atMost(70, TimeUnit.SECONDS)
        .untilAsserted(() -> {
            MessageJsonRepresentation jsonPresentation =
                RestAssured.given()
                    .get(JSON_SERVER_EXPOSED + "/messages/" + id)
                    .then()
                    .extract()
                    .as(MessageJsonRepresentation.class);

            assertNotNull(jsonPresentation);
        });
}
```

### Conclusion

> Hopefully, the described approaches and code examples will help make it easier to test kafka clients application. If you would like to contribute, just open a ticket on Github for ideas. 

Reviewer 2: ```suggestion
Hopefully, the described approaches and code examples will help to make testing Kafka client application easier. If you would like to contribute, just open a ticket on GitHub for ideas. 
```

> Additionally, I would recommend to check Confluent blog that can be helpful in learning more about Kafka.

Reviewer 2: ```suggestion
Additionally, I would recommend to check [Confluent's blog](https://www.confluent.io/blog/) that can be helpful to learn more about Apache Kafka.
```
